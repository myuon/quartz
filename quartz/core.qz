// @[nostd]
let error_not_an_pointer = 1;

let error_stack_overflow = 2;

let error_stack_underflow = 3;

let error_nil_pointer_exception = 4;

let error_invalid_block = 5;

let error_out_of_memory = 6;

let error_invalid_type = 7;

let error_type_reps_ptr_not_initialized = 8;

let error_invalid_header = 9;

// 1.5 * 1024 * 1024 * 1024
let stack_limit: u32 = 1610612736 as u32;

let _sp: u32 = 1610612736 as u32;

let _bp: u32 = 1610612736 as u32;

@[no_allocation]
fun push_memstack(value: ptr[any]) {
    // FIXME: Removing the following comment will cause check_if_stable to fail.
    // if !_reflection_is_pointer(value) {
    //     debug(error_not_an_pointer);
    //     debug(value);
    //     abort();
    // }

    let sp_ptr = _sp as ptr[ptr[byte]];
    sp_ptr.at(0) = value;

    _sp = _sp - (sizeof[ptr[any]]() as u32);

    if (_sp - (32 as u32)) < memory_block_limit {
        debug(error_stack_overflow);
        abort();
    }
}

@[no_allocation]
fun prologue() {
    push_memstack(_bp as ptr[any]);

    _bp = _sp;
}

@[no_allocation]
fun epilogue() {
    _sp = _bp;

    let bp_ptr = (_bp as ptr[ptr[byte]]).offset(1);
    _bp = bp_ptr.at(0) as u32;

    // FIXME: when _bp == 0?
    if _bp != (0 as u32) && (_bp < _sp || _bp > stack_limit) {
        debug(_bp);
        debug(stack_limit);
        debug(error_stack_underflow);
        abort();
    }
}

@[no_allocation]
fun validate_address(address: i32) {
    let p = address as ptr[any];
    if p == nil as ptr[any] {
        debug(error_nil_pointer_exception);
        debug(1);
        abort();
    }
}

let memory_block_limit: u32 = 1073741824 as u32;

let strings_ptr: u32 = 0 as u32;

@[no_allocation]
fun load_string(index: i32): string {
    if strings_ptr == (0 as u32) {
        debug(error_nil_pointer_exception);
        debug(2);
        abort();
    }

    return (strings_ptr as ptr[string]).at(index);
}

let alloc_ptr: u32 = 0 as u32;

let alloc_count = 0;

@[no_allocation]
fun alloc_with_rep(rep: ptr[ptr[byte]], size: i32): ptr[any] {
    let size_ = size;
    if size % sizeof[i32]() != 0 {
        size_ = ((size + sizeof[i32]() - 1) / sizeof[i32]()) * sizeof[i32]();
    }
    if size == 0 {
        return nil as ptr[any];
    }

    let current = alloc_ptr;
    alloc_ptr = alloc_ptr + (size_ as u32);

    let block = find_free_block((size_ as u32) + Header::sizeof());
    if !block.get_is_free() {
        debug(error_invalid_block);
        abort();
    }

    // split
    block.set_size(block.get_size() - (size_ as u32) - Header::sizeof());
    if block.get_size() <= (0 as u32) {
        debug(error_invalid_block);
        abort();
    }
    block.set_is_free(true);

    let current = block.get_next();
    current.set_size(size_ as u32);
    current.set_is_free(false);
    current.set_type_rep(rep);

    alloc_count = alloc_count + 1;
    if alloc_count > 5000000 {
        perform_gc();

        alloc_count = 0;
    }

    return current.get_data_ptr();
}

@[no_allocation]
fun find_free_block(size: u32): Header {
    if (heap_root_ptr as ptr[any]) == (nil as ptr[any]) {
        let current = Header::from_ptr(alloc_ptr as ptr[any]);
        current.set_size(memory_block_limit);
        current.set_is_free(true);
        heap_root_ptr = alloc_ptr;
    }

    let current = Header::from_ptr(heap_root_ptr as ptr[any]);
    while (current as u32) < (_sp - (32 as u32)) {
        if current.get_is_free() && current.get_size() > size {
            return current;
        }

        current = current.get_next();
    }

    // try again!
    perform_gc();

    let current = Header::from_ptr(heap_root_ptr as ptr[any]);
    while (current as u32) < (_sp - (32 as u32)) {
        if current.get_is_free() && current.get_size() > size {
            return current;
        }

        current = current.get_next();
    }

    debug(error_out_of_memory);
    abort();
}

// size: u32,
// header_flags: u32,
// rep: *StructData[TypeRep]
struct Header {
}

module Header {
    @[no_allocation]
    fun sizeof(): u32 {
        return (sizeof[u32]() + sizeof[u32]() + sizeof[ptr[byte]]()) as u32;
    }

    @[no_allocation]
    fun from_ptr(p: ptr[byte]): Header {
        return p as Header;
    }

    @[no_allocation]
    fun from_data_ptr(p: ptr[byte]): Header {
        let header = Header::from_ptr(p.offset(0 - (Header::sizeof() as i32)));
        header.validate();

        return header;
    }

    @[no_allocation]
    fun get_next(self): Header {
        return Header::from_ptr((self as ptr[byte]).offset((Header::sizeof() + self.get_size()) as i32));
    }

    @[no_allocation]
    fun get_data_ptr(self): ptr[byte] {
        return (self as ptr[byte]).offset(Header::sizeof() as i32);
    }

    @[no_allocation]
    fun get_is_free(self): bool {
        return (self.get_header_flags() & 1 as u32) == (1 as u32);
    }

    @[no_allocation]
    fun set_is_free(self, value: bool) {
        if value {
            self.set_header_flags(self.get_header_flags() | (1 as u32));
        } else {
            self.set_header_flags(self.get_header_flags() & bit_not_u32(1 as u32));
        }
    }

    @[no_allocation]
    fun get_is_marked(self): bool {
        return (self.get_header_flags() & 2 as u32) == 2 as u32;
    }

    @[no_allocation]
    fun set_is_marked(self, value: bool) {
        if value {
            self.set_header_flags(self.get_header_flags() | (2 as u32));
        } else {
            self.set_header_flags(self.get_header_flags() & bit_not_u32(2 as u32));
        }
    }

    @[no_allocation]
    fun get_size(self): u32 {
        return (self as ptr[any]).at(0);
    }

    @[no_allocation]
    fun set_size(self, value: u32) {
        (self as ptr[any]).at(0) = value;
    }

    @[no_allocation]
    fun get_header_flags(self): u32 {
        return (self as ptr[any]).at(1);
    }

    @[no_allocation]
    fun set_header_flags(self, value: u32) {
        (self as ptr[any]).at(1) = value;
    }

    @[no_allocation]
    fun get_type_rep(self): ptr[ptr[byte]] {
        return (self as ptr[any]).at(2);
    }

    @[no_allocation]
    fun set_type_rep(self, value: ptr[ptr[byte]]) {
        (self as ptr[any]).at(2) = value;
    }

    @[no_allocation]
    fun validate(self) {
        if !types::is_i32(self.get_size()) {
            debug(error_invalid_header);
            debug(1);
            debug(self);
            abort();
        }

        if !types::is_i32(self.get_header_flags()) {
            debug(error_invalid_header);
            debug(2);
            debug(self);
            abort();
        }

        if !types::is_pointer(self.get_type_rep()) {
            debug(error_invalid_header);
            debug(3);
            debug(self);
            abort();
        }

        if !is_typerep_pointer(self.get_type_rep() as ptr[byte]) {
            debug(error_invalid_header);
            debug(4);
            debug(self);
            abort();
        }
    }
}

@[no_allocation]
fun bit_not_u32(x: u32): u32 {
    return xor_u32(x, 4294967295); // 2^32 - 1
}

fun new_string(p: ptr[byte], length: i32): string {
    return string {
        data: p,
        length: length,
    };
}

let type_reps_ptr: ptr[ptr[any]] = (nil as ptr[ptr[any]]);

@[no_allocation]
fun get_type_rep_address(rep_id: i32): ptr[ptr[any]] {
    if type_reps_ptr == (nil as ptr[ptr[any]]) {
        debug(error_type_reps_ptr_not_initialized);
        abort();
    }

    return type_reps_ptr.offset(rep_id);
}

@[no_allocation]
fun get_typetag(value: any): i32 {
    return _get_typetag(value);
}

@[no_allocation]
fun is_stack_pointer(value: ptr[byte]): bool {
    return (value as u32) >= _sp;
}

@[no_allocation]
fun is_typerep_pointer(p: ptr[byte]): bool {
    let p_address = p as i32;
    let type_reps_ptr_address = type_reps_ptr as i32;

    return p_address >= type_reps_ptr_address && p_address < ((type_reps_ptr as ptr[any]).offset(
        type_reps_count,
    )) as i32;
}

