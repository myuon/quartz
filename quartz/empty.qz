struct Foo {
  x: i32,
}

fun main() {
  let f = Foo {x: 42};
  let x = 10;

  x.to_string()
}
