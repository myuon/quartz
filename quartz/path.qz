struct Path {
    path: vec[string],
}

module Path {
    fun new(): Path {
        return Path {
            path: make[vec[string]](),
        };
    }

    fun push(self, s: string) {
        self.path.push(s);
    }

    fun ident(i: string): Path {
        return Path {
            path: make[vec[string]](i),
        };
    }

    fun join(self, delimiter: string): string {
        if self.path.length == 0 {
            return "";
        } else {
            let first = self.path.at(0);

            for i in 1..self.path.length {
                first = first.concat(delimiter).concat(self.path.at(i));
            }

            return first;
        }
    }
}

