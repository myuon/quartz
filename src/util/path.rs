use super::ident::Ident;

pub struct Path(pub Vec<Ident>);

impl Path {
    pub fn new(segments: Vec<Ident>) -> Path {
        Path(segments)
    }

    pub fn ident(ident: Ident) -> Path {
        Path(vec![ident])
    }
}
