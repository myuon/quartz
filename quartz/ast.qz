import quartz::ast;
import quartz::path;
import quartz::location;

struct Ident {
    data: string,
    location: Location,
}

struct Module {
    decls: vec[LDecl],
}

module Module {
    fun to_string(self): string {
        return derive::to_string(self);
    }
}

enum Decl {
    t_func: Function,
    t_let: Let,
    t_type: struct {
        name: string,
        type_: Type,
    },
    t_enum: struct {
        name: string,
        type_: Type,
    },
    t_module: struct {
        path: Path,
        extends: string?,
        module_: Module,
    },
    t_import: Path,
    t_attributes: LAttributes,
    t_skip: bool,
    t_declare_func: DeclareFunc,
    t_interface: Interface,
}

module Decl {
    fun to_string(self): string {
        return derive::to_string(self);
    }
}

struct Interface {
    name: string,
    methods: vec[struct {
        name: string,
        params: vec[struct {
            name: string,
            type_: Type,
        }],
        result_type: Type,
    }],
}

struct DeclareFunc {
    name: Ident,
    result_type: Type,
    params: vec[struct {
        name: string,
        type_: Type,
    }],
    declare_namespace: string?,
    declare_wrap: struct {
        params: vec[Type],
        result_type: Type,
    }?,
}

module DeclareFunc {
    fun to_type(self): Type {
        let params_type = make[vec[Type]]();
        for i in 0..self.params.length {
            let p = self.params.at(i);
            params_type.push(p.type_);
        }

        return Type {
            t_func: struct {
                params: params_type,
                result: self.result_type,
            },
        };
    }
}

struct LDecl {
    data: Decl,
    location: Location,
}

struct Let {
    name: string,
    name_type: Type,
    expr: LExpression,
}

module Let {
    fun to_string(self): string {
        return derive::to_string(self);
    }
}

struct Function {
    name: Ident,
    result_type: Type,
    body: LBlock,
    params: vec[struct {
        name: string,
        type_: Type,
    }],
    variadic: struct {
        name: string,
        type_: Type,
    }?,
    no_allocation: bool,
    is_test: bool,
    ffi_export: string?,
}

module Function {
    fun to_string(self): string {
        return derive::to_string(self);
    }

    fun to_type(self): Type {
        let params_type = make[vec[Type]]();
        for i in 0..self.params.length {
            let p = self.params.at(i);
            params_type.push(p.type_);
        }

        if self.variadic != nil {
            return Type {
                t_variadic_func: struct {
                    params: params_type,
                    result: self.result_type,
                    variadic: self.variadic!.type_,
                },
            };
        } else {
            return Type {
                t_func: struct {
                    params: params_type,
                    result: self.result_type,
                },
            };
        }
    }
}

struct Block {
    block: vec[LStatement],
}

module Block {
    fun to_string(self): string {
        return derive::to_string(self);
    }
}

struct LBlock {
    data: Block,
    location: Location,
}

enum Statement {
    t_let: struct {
        pattern: LPattern,
        name_type: Type,
        expr: LExpression,
    },
    t_return: struct {
        expr: LExpression,
    },
    t_assign: struct {
        lhs: LExpression,
        rhs_type: Type?,
        rhs: LExpression,
    },
    t_expr: struct {
        expr: LExpression,
    },
    t_if: struct {
        condition: LExpression,
        then_block: LBlock,
        else_block: LBlock?,
    },
    t_while: struct {
        condition: LExpression,
        block: LBlock,
    },
    t_for: struct {
        mode: string,
        type_: Type?,
        ident: string,
        range: LExpression,
        block: LBlock,
    },
    t_continue: bool,
    t_break: bool,
    t_attribute: LAttribute,
}

module Statement {
    fun to_string(self): string {
        return derive::to_string(self);
    }
}

struct LStatement {
    data: Statement,
    location: Location,
}

enum Expression {
    t_ident: struct {
        name: string,
        resolved_path: Path?,
    },
    t_binop: struct {
        binop: string,
        type_: Type?,
        left: LExpression,
        right: LExpression,
    },
    t_i32: struct {
        value: i32,
    },
    t_u32: struct {
        value: u32,
    },
    t_call: struct {
        callee: LExpression,
        args: vec[LExpression],
        variadic: struct {
            element_type: Type,
            index: i32,
        }?,
        expansion: LExpression?,
    },
    t_project: struct {
        expr: LExpression,
        field: Ident,
        type_: Type?,
        resolved_path: Path?,
    },
    t_record: RecordExpression,
    t_anonymous_record: struct {
        fields: vec[struct {
            data: struct {
                field: string,
                value: LExpression,
            },
            location: Location,
        }],
        type_: Type?,
    },
    t_make: struct {
        type_: Type,
        args: vec[LExpression],
    },
    t_as: struct {
        expr: LExpression,
        source_type: Type,
        target_type: Type,
    },
    t_path: struct {
        path: Path,
        resolved_path: Path?,
    },
    t_self: bool,
    t_string: struct {
        value: string,
        raw: string,
    },
    t_paren: LExpression,
    t_range: struct {
        start: LExpression,
        end: LExpression,
    },
    t_unop: struct {
        unop: string,
        expr: LExpression,
    },
    t_unwrap: struct {
        type_: Type,
        expr: LExpression,
    },
    t_wrap: struct {
        type_: Type,
        expr: LExpression,
    },
    t_bool: bool,
    t_nil: bool, // if nil is explicit
    t_omit: bool,
    t_enumor: struct {
        lhs: LExpression?,
        rhs: LExpression?,
        lhs_type: Type,
        rhs_type: Type,
    },
    t_sizeof: struct {
        type_: Type,
    },
    t_try: LExpression,
    t_upcast: struct {
        interface_name: string,
        expr: LExpression,
    },
}

module Expression {
    fun to_string(self): string {
        return derive::to_string(self);
    }
}

struct LExpression {
    data: Expression,
    location: Location,
}

struct RecordExpression {
    name: string,
    fields: vec[struct {
        field: string,
        value: LExpression,
    }],
    expansion: LExpression?,
}

module RecordExpression {
    fun to_string(self): string {
        return derive::to_string(self);
    }

    fun at(self, field: string): Expression or error {
        for i in 0..self.fields.length {
            let f = self.fields.at(i);
            if f.field == field {
                return f.value.data;
            }
        }

        return _ or error::new("Field not found");
    }

    fun to_fields_map(self): map[string, LExpression] {
        let result = make[map[string, LExpression]]();
        for i in 0..self.fields.length {
            let f = self.fields.at(i);
            result.insert(f.field, f.value);
        }

        return result;
    }
}

enum Type {
    t_omit: bool, // for unification
    t_nil: bool,
    t_i32: bool,
    t_u32: bool,
    t_func: struct {
        params: vec[Type],
        result: Type,
    },
    t_variadic_func: struct {
        params: vec[Type],
        result: Type,
        variadic: Type,
    },
    t_ptr: Type,
    t_bool: bool,
    t_struct: TypeStruct,
    t_byte: bool,
    t_ident: string,
    t_range: Type,
    t_optional: Type,
    t_vec: Type,
    t_or: struct {
        left: Type,
        right: Type,
    },
    t_any: bool,
    t_map: struct {
        key: Type,
        value: Type,
    },
    t_enum: TypeStruct,
}

module Type {
    fun to_string(self): string {
        if self.t_omit != nil {
            return "_";
        } else if self.t_nil != nil {
            return "nil";
        } else if self.t_i32 != nil {
            return "i32";
        } else if self.t_u32 != nil {
            return "u32";
        } else if self.t_bool != nil {
            return "bool";
        } else if self.t_byte != nil {
            return "byte";
        } else if self.t_any != nil {
            return "any";
        } else if self.t_ident != nil {
            return self.t_ident!;
        } else if self.t_range != nil {
            return "range[{}]".format(self.t_range!.to_string());
        } else if self.t_optional != nil {
            return "{}?".format(self.t_optional!.to_string());
        } else if self.t_vec != nil {
            return "vec[{}]".format(self.t_vec!.to_string());
        } else if self.t_or != nil {
            return "{} or {}".format(self.t_or!.left.to_string(), self.t_or!.right.to_string());
        } else if self.t_map != nil {
            return "map[{}, {}]".format(self.t_map!.key.to_string(), self.t_map!.value.to_string());
        } else if self.t_struct != nil {
            let builder = stringbuilder::new();
            builder.append("struct {");
            for i in 0..self.t_struct!.fields.length {
                let field = self.t_struct!.fields.at(i);
                builder.append(format("{}: {}", field.data.name, field.data.type_.to_string()));

                if i != self.t_struct!.fields.length - 1 {
                    builder.append(", ");
                }
            }
            builder.append("}");

            return builder.to_string();
        } else if self.t_enum != nil {
            let builder = stringbuilder::new();
            builder.append("enum {");
            for i in 0..self.t_enum!.fields.length {
                let field = self.t_enum!.fields.at(i);
                builder.append(field.data.name);

                if i != self.t_enum!.fields.length - 1 {
                    builder.append(", ");
                }
            }
            builder.append("}");

            return builder.to_string();
        } else if self.t_func != nil {
            let builder = stringbuilder::new();
            builder.append("fun (");
            for i in 0..self.t_func!.params.length {
                let param = self.t_func!.params.at(i);
                builder.append(param.to_string());

                if i != self.t_func!.params.length - 1 {
                    builder.append(", ");
                }
            }
            builder.append(format(") -> {}", self.t_func!.result.to_string()));

            return builder.to_string();
        } else if self.t_variadic_func != nil {
            let builder = stringbuilder::new();
            builder.append("fun (");
            for i in 0..self.t_variadic_func!.params.length {
                let param = self.t_variadic_func!.params.at(i);
                builder.append(param.to_string());

                if i != self.t_variadic_func!.params.length - 1 {
                    builder.append(", ");
                }
            }
            builder.append(format(
                ", ..{}) -> {}",
                self.t_variadic_func!.variadic.to_string(),
                self.t_variadic_func!.result.to_string(),
            ));

            return builder.to_string();
        } else if self.t_ptr != nil {
            return "ptr[{}]".format(self.t_ptr!.to_string());
        }

        panic("Unknown type: {}", derive::to_string(self));
        return "";
    }

    fun string(): Type {
        return Type {
            t_ident: "string",
        };
    }

    fun is_integer_type(self): bool {
        return self.t_i32 != nil || self.t_u32 != nil || (self.t_ident != nil && self.t_ident!.equal("i64"));
    }
}

struct TypeStruct {
    fields: vec[struct {
        data: struct {
            name: string,
            type_: Type,
        },
        location: Location,
    }],
}

module TypeStruct {
    fun has(self, name: string): bool {
        for i in 0..self.fields.length {
            if self.fields.at(i).data.name.equal(name) {
                return true;
            }
        }
        return false;
    }

    fun at(self, name: string): Type {
        for i in 0..self.fields.length {
            if self.fields.at(i).data.name.equal(name) {
                return self.fields.at(i).data.type_;
            }
        }

        return panic("unimplemented: typestruct.at");
    }

    fun index(self, name: string): i32 {
        for i in 0..self.fields.length {
            if self.fields.at(i).data.name.equal(name) {
                return i;
            }
        }

        return panic("unimplemented: typestruct.index, {}, {}".format(derive::to_string(self), name));
    }
}

struct LType {
    data: Type,
    location: Location,
}

struct Pattern {
    t_ident: string?,
    t_or: struct {
        lhs: LPattern,
        rhs: LPattern,
    }?,
    t_omit: bool?,
}

module Pattern {
    fun to_string(self): string {
        return derive::to_string(self);
    }
}

struct LPattern {
    data: Pattern,
    location: Location,
}

struct Attribute {
    name: Ident,
    args: vec[LExpression],
}

struct LAttribute {
    data: Attribute,
    location: Location,
}

struct LAttributes {
    data: vec[Attribute],
    location: Location,
}

