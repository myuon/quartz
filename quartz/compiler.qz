import quartz::lexer;
import quartz::parser;
import quartz::typecheck;
import quartz::ir_code_gen;
import quartz::generator;
import quartz::path;
import quartz::ast;
import quartz::ir;
import quartz::value;
import quartz::location;
import quartz::preprocessor;
import quartz::ir_path::let_call;

struct LoadedModule {
    path: Path,
    source: string,
    module_: Module,
}

struct Compiler {
    lib_path_string: string?,
    loaded: vec[LoadedModule],
    version: string,
}

module Compiler {
    fun new(lib_path_string: string?, version: string): Compiler {
        return Compiler {
            lib_path_string: lib_path_string,
            loaded: make[vec[LoadedModule]](),
            version: version,
        };
    }

    fun load_module(self, path: Path, skip_errors: bool, preprocess: bool): vec[Path] or error {
        if self.lib_path_string != nil {
            let path_string = format("{}/{}.qz", self.lib_path_string!, path.join("/"));
            if file_exists(path_string).try {
                let input = file_read(path_string).try;

                return self.register_module(path, input, skip_errors, preprocess).try;
            }
        }

        let input = file_read(format("{}.qz", path.join("/"))).try;

        return self.register_module(path, input, skip_errors, preprocess).try;
    }

    fun matches(self, path: Path): LoadedModule? {
        for loaded in self.loaded {
            if is_prefix_path(path, loaded.path) {
                return loaded?;
            }
        }

        return nil;
    }

    fun register_module(
        self,
        path: Path,
        input: string,
        skip_errors: bool,
        preprocess: bool,
    ): vec[Path] or error {
        let lexer = Lexer::new();
        lexer.run(input).try;

        let parser = Parser::new();
        let module_ = parser.run(lexer.tokens, skip_errors).try;

        if preprocess {
            let preprocessor = Preprocessor::new();
            preprocessor.run(
                module_,
                struct {
                    version: self.version,
                },
            ).try;
        }

        self.loaded.push(LoadedModule {
            path: path,
            source: input,
            module_: module_,
        });

        return parser.imports;
    }

    fun parse(
        self,
        input: string,
        load_path: Path?,
        skip_errors: bool,
        preprocess: bool,
    ): Module or error {
        let main_path = Path::new("main");
        if load_path != nil {
            main_path = load_path!;
        }
        let imports = self.register_module(main_path, input, skip_errors, preprocess).try;

        let visited = make[map[string, bool]]();
        visited.insert(main_path.join("_"), true);
        if !input.starts_with("// @[nostd]") {
            imports.push(Path::new("quartz", "std"));
        }

        while imports.length != 0 {
            let path = vec_pop(imports) as Path;
            if visited.has(path.join("_")) {
                continue;
            }

            imports.extend(self.load_module(path, skip_errors, preprocess).try);
            visited.insert(path.join("_"), true);
        }

        let decls = make[vec[LDecl]]();
        for loaded in self.loaded {
            decls.push(LDecl {
                data: Decl {
                    t_module: struct {
                        path: loaded.path,
                        extends: nil,
                        module_: loaded.module_,
                    },
                },
                location: Location::unknown(),
            });
        }

        let module_ = Module {
            decls: decls,
        };

        return module_;
    }

    fun compile(
        self,
        input: string,
        main_path: Path,
        entrypoint: string,
        skip_errors: bool,
        validate_address: bool,
        test_mode: bool,
        output_path: string,
    ): nil or error {
        let module_ = self.parse(input, nil, skip_errors, true).context("parse").try;

        let typechecker = Typechecker::new();
        let n or err = typechecker.run(module_, main_path).context("typecheck");
        if err != nil {
            let err_internal = err!;

            for detail in err!.detail {
                let detail_type = reflection::get_type_rep(detail);
                if detail_type.name.equal("ErrorSource") {
                    let source = detail as ErrorSource;

                    let input_to_inspect = input;
                    if main_path.join("_") != source.path.join("_") {
                        let loaded = self.matches(source.path);
                        if loaded == nil {
                            return _ or err!;
                        }

                        input_to_inspect = loaded!.source;
                    }

                    let line_counts = input_to_inspect.line_counts();

                    let start_position = find_position_with_line_counts(line_counts, source.location.start!);
                    let end_position = find_position_with_line_counts(line_counts, source.location.end!);

                    err_internal = err_internal.context(struct {
                        path: source.path,
                        start: start_position,
                        end: end_position,
                    });
                }
            }

            return _ or err_internal;
        }

        let irgen = IrCodeGenerator::new(test_mode, "main_{}".format(entrypoint));
        irgen.globals = typechecker.globals;
        irgen.interfaces = typechecker.interfaces;
        let term = irgen.run(module_).context("ircodegen").try;

        term = transform_let_call(term);

        let envs = environs();
        if envs.has("GENERATE_IR") {
            file_write("quartz.ir", term.to_string());

            println("wrote quartz.ir");
        }

        let writer = FileWriter::open(output_path).try;

        let gen = Generator::new(writer, validate_address);
        gen.set_globals(typechecker.globals);
        gen.set_strings(irgen.strings.strings);
        gen.set_type_reps(irgen.type_reps.type_reps);
        gen.set_start_terms(irgen.start);

        term = gen.fold_consts(term);

        let offset = ((irgen.data_section_offset + 7) / 8) * 8;
        gen.run(term, entrypoint, offset);

        writer.close();

        return nil;
    }

    fun check(
        self,
        input: string,
        path: Path,
    ): vec[struct {
        start: struct {
            line: i32,
            column: i32,
        },
        end: struct {
            line: i32,
            column: i32,
        },
        message: string,
    }] {
        let module_ or err = self.parse(input, nil, false, false);
        if err != nil {
            let err_internal = err!;

            for detail in err_internal.detail {
                let detail_type = reflection::get_type_rep(detail);
                if detail_type.name.equal("ErrorSource") {
                    let source = detail as ErrorSource;

                    if source.location.start == nil || source.location.end == nil {
                        panic("NPE");
                    }

                    let start_position = find_position(input, source.location.start!);
                    let end_position = find_position(input, source.location.end!);

                    return make[vec[struct {
                        start: struct {
                            line: i32,
                            column: i32,
                        },
                        end: struct {
                            line: i32,
                            column: i32,
                        },
                        message: string,
                    }]](struct {
                        start: start_position,
                        end: end_position,
                        message: err_internal.message,
                    });
                }
            }

            return make[vec[struct {
                start: struct {
                    line: i32,
                    column: i32,
                },
                end: struct {
                    line: i32,
                    column: i32,
                },
                message: string,
            }]](struct {
                start: struct {
                    line: 0,
                    column: 0,
                },
                end: struct {
                    line: 0,
                    column: 0,
                },
                message: format("parse error: {}", err!.message),
            });
        }

        let typechecker = Typechecker::new();
        let n or err = typechecker.run(module_!, path);
        if err == nil {
            return make[vec[struct {
                start: struct {
                    line: i32,
                    column: i32,
                },
                end: struct {
                    line: i32,
                    column: i32,
                },
                message: string,
            }]]();
        }

        let err_internal = err!;

        for detail in err_internal.detail {
            let detail_type = reflection::get_type_rep(detail);
            if detail_type.name.equal("ErrorSource") {
                let source = detail as ErrorSource;

                if source.location.start == nil || source.location.end == nil {
                    panic("NPE");
                }

                let start_position = find_position(input, source.location.start!);
                let end_position = find_position(input, source.location.end!);

                return make[vec[struct {
                    start: struct {
                        line: i32,
                        column: i32,
                    },
                    end: struct {
                        line: i32,
                        column: i32,
                    },
                    message: string,
                }]](struct {
                    start: start_position,
                    end: end_position,
                    message: err_internal.message,
                });
            }
        }

        return make[vec[struct {
            start: struct {
                line: i32,
                column: i32,
            },
            end: struct {
                line: i32,
                column: i32,
            },
            message: string,
        }]](struct {
            start: struct {
                line: 0,
                column: 0,
            },
            end: struct {
                line: 0,
                column: 0,
            },
            message: format("typecheck error: {}", err_internal.message),
        });
    }

    fun check_type(self, filepath: string, input: string, line: i32, column: i32): string or error {
        let path = Path::new("main");

        let module_ = self.parse(input, nil, true, false).try;

        let position = find_line_column_from_position(input, line, column);

        let typechecker = Typechecker::new();
        let t = typechecker.find_at_cursor(module_, path, position).try;

        return format("```quartz\n{}\n```", t.to_string());
    }

    fun completion(
        self,
        input: string,
        path: Path,
        line: i32,
        column: i32,
        dot: bool,
    ): vec[struct {
        kind: string,
        label: string,
        text: string,
    }] or error {
        let module_ = self.parse(input, nil, true, false).try;
        let position = find_line_column_from_position(input, line, column);

        let typechecker = Typechecker::new();
        let result = typechecker.completion(module_, path, position, dot).try;

        return result;
    }

    fun go_to_def(
        self,
        input: string,
        path: Path,
        line: i32,
        column: i32,
    ): struct {
        module_name: string,
        start: struct {
            line: i32,
            column: i32,
        },
        end: struct {
            line: i32,
            column: i32,
        },
    } or error {
        let module_ = self.parse(input, path?, true, false).try;
        let position = find_line_column_from_position(input, line, column);

        let typechecker = Typechecker::new();
        let result = typechecker.find_definition(module_, path, position).try;

        let loaded = self.matches(path);
        if loaded != nil {
            return struct {
                module_name: loaded!.path.join("/"),
                start: find_position(loaded!.source, result.start),
                end: find_position(loaded!.source, result.end),
            };
        } else {
            return _ or error::new("module not found");
        }
    }
}

fun find_position(
    input: string,
    position: i32,
): struct {
    line: i32,
    column: i32,
} {
    return find_position_with_line_counts(input.line_counts(), position);
}

fun find_position_with_line_counts(
    line_counts: vec[i32],
    position: i32,
): struct {
    line: i32,
    column: i32,
} {
    let line_number = 0;
    let count = 0;
    for line in line_counts {
        if count + line + 1 > position {
            break;
        }

        line_number = line_number + 1;
        count = count + line + 1;
    }

    return struct {
        line: line_number,
        column: position - count,
    };
}

fun find_line_column_from_position(input: string, line: i32, column: i32): i32 {
    let start = 0;
    let end = input.length;
    let line_counts = input.line_counts();

    while start < end {
        let mid = (start + end) / 2;
        let pos = find_position_by_line_counts(line_counts, mid);

        if pos.line == line && pos.column == column {
            return mid;
        }

        if pos.line < line || (pos.line == line && pos.column < column) {
            start = mid + 1;
        } else {
            end = mid;
        }
    }

    return start;
}

fun find_position_by_line_counts(
    line_counts: vec[i32],
    position: i32,
): struct {
    line: i32,
    column: i32,
} {
    let line_number = 0;
    let count = 0;
    for line in line_counts {
        if count + line > position {
            break;
        }

        line_number = line_number + 1;
        count = count + line + 1;
    }

    return struct {
        line: line_number,
        column: position - count,
    };
}

