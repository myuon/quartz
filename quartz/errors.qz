import quartz::path;
import quartz::location;

struct ErrorSource {
    path: Path,
    location: Location,
}

