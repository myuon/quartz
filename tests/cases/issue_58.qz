import quartz::path;

fun main() {
    Path::new(..path.path.slice(0, 2));
}

