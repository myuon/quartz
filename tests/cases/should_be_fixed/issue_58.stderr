project.resolved_path is nil: main